# Examen

## Datos

Alumno

**Nombre:** Alan Ruiz Mondragón.  
**Grupo:** 18.

## Librerias

In [458]:
#Importamos librerias
import datetime
import emoji
import jellyfish as jf
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import re 
import seaborn as sns
import unicodedata

#Definimos configuraciones
pd.set_option("display.max_columns",200)
pd.set_option("display.max_rows",200)

## Funciones

In [459]:
#Definimos una función para limpieza de texto
def clean_text(text, pattern="[^a-zA-Z0-9]"):
    text=str(text)
    cleaned_text = unicodedata.normalize('NFD', text).encode('ascii', 'ignore')
    cleaned_text = re.sub(pattern, " ", cleaned_text.decode("utf-8"), flags=re.UNICODE)
    cleaned_text = u' '.join(cleaned_text.lower().strip().lstrip().split())
    return cleaned_text if cleaned_text!="nan" else np.nan

#Definimos una función para revisar la completitud de nuestras variables
def completitud(df):
    comple=pd.DataFrame(df.isnull().sum())
    comple.reset_index(inplace=True)
    comple=comple.rename(columns={"index":"columna",0:"total"})
    comple["completitud"]=(1-comple["total"]/df.shape[0])*100
    comple=comple.sort_values(by="completitud",ascending=True)
    comple.reset_index(drop=True,inplace=True)
    return comple

## Datasets

In [460]:
#Cargamos los datos
#Cargamos el dataset principal
df = pd.read_csv('/Users/alanruizmondragon/Documents/PERSONAL/Code/Diplomados/Ciencia de datos/Modulo I/examen/data/data_examen.csv')
#Nombramos al dataset secuendario
df1 = pd.read_csv('/Users/alanruizmondragon/Documents/PERSONAL/Code/Diplomados/Ciencia de datos/Modulo I/examen/data/ocupaciones_examen.csv')

### Créditos

**Contenido**

El primer dataset hace referencia información relacionada con las características de personas que aplicaron para solicitar un crédito, adicional a esa información se cuenta
con una variable ”tgt” que representa una variable objetivo que toma el valor de 1 si se otorgó el crédito y 0 si se rechazó la solicitud.

In [461]:
#Visualizamos el dataset principal
df.head(5)

ID_CLIENT  ID_SHOP SEX MARITAL_STATUS   AGE  QUANT_DEPENDANTS  EDUCATION  \
0          2       15   F              S  18.0                 0        NaN   
1          4       12   F              C   NaN                 0        NaN   
2          5       16   F              S  28.0                 0        NaN   
3          6       24   M              S  26.0                 0        NaN   
4          7       55   F              S  22.0                 0        NaN   

  FLAG_RESIDENCIAL_PHONE  AREA_CODE_RESIDENCIAL_PHONE  PAYMENT_DAY  SHOP_RANK  \
0                      Y                           31           20          0   
1                      N                           31           25          0   
2                      Y                           31           25          0   
3                      N                           31           28          0   
4                      Y                           31           12          0   

  RESIDENCE_TYPE  MONTHS_IN_RESIDENCE FLAG_MOTHERS_NAME FLAG_FATHERS_NAME  \
0              P                  216                 Y                 Y   
1              P                  180                 Y                 Y   
2                                  12                 Y                 Y   
3              P                  180                 Y                 Y   
4              A                    0                 Y                 Y   

  FLAG_RESIDENCE_TOWN=WORKING_TOWN FLAG_RESIDENCE_STATE=WORKING_STATE  \
0                                Y                                  Y   
1                                N                                  Y   
2                                Y                                  Y   
3                                N                                  Y   
4                                Y                                  Y   

   MONTHS_IN_THE_JOB  PROFESSION_CODE  MATE_INCOME  \
0                 12              853          0.0   
1                 24               35          0.0   
2                 12               24          0.0   
3                  0              999          0.0   
4                 48              999          0.0   

  FLAG_RESIDENCIAL_ADDRESS=POSTAL_ADDRESS FLAG_OTHER_CARD  \
0                                       Y               N   
1                                       Y               N   
2                                       Y               N   
3                                       Y               N   
4                                       Y               N   

   QUANT_BANKING_ACCOUNTS PERSONAL_REFERENCE_#1      PERSONAL_REFERENCE_#2  \
0                       0                  SARA                     FELIPE   
1                       0                  JACI  VALERIA ALEXANDRA TRAJANO   
2                       0                   NaN         SANDRO L P MARTINS   
3                       0                   NaN                        ANA   
4                       0                   NaN                        NaN   

  FLAG_MOBILE_PHONE FLAG_CONTACT_PHONE PERSONAL_NET_INCOME  \
0                 N                  N               300.0   
1                 N                  N               304.0   
2                 N                  N               250.0   
3                 N                  N               800.0   
4                 N                  N               410.0   

   COD_APPLICATION_BOOTH  QUANT_ADDITIONAL_CARDS_IN_THE_APPLICATION  \
0                      0                                          0   
1                      0                                          0   
2                      0                                          0   
3                      0                                          0   
4                      0                                          0   

  FLAG_CARD_INSURANCE_OPTION  tgt  
0                          N    0  
1                          N    0  
2                          N    0  
3                          N    0 

**Tamaño**

In [462]:
#Visualizamos el tamaño del dataset
df.shape

(51000, 32)

**Formato**

In [463]:
#Visualizamos la información del dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51000 entries, 0 to 50999
Data columns (total 32 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   ID_CLIENT                                  51000 non-null  int64  
 1   ID_SHOP                                    51000 non-null  int64  
 2   SEX                                        50997 non-null  object 
 3   MARITAL_STATUS                             51000 non-null  object 
 4   AGE                                        48528 non-null  float64
 5   QUANT_DEPENDANTS                           51000 non-null  int64  
 6   EDUCATION                                  0 non-null      float64
 7   FLAG_RESIDENCIAL_PHONE                     51000 non-null  object 
 8   AREA_CODE_RESIDENCIAL_PHONE                51000 non-null  int64  
 9   PAYMENT_DAY                                51000 non-null  int64  
 10  SHOP_RANK             

**Diccionario**

In [464]:
#Cargamos el diccionario de datos del dataset principal
diccionario = pd.read_excel('/Users/alanruizmondragon/Documents/PERSONAL/Code/Diplomados/Ciencia de datos/Modulo I/examen/diccionario/VariablesList.xlsx')

In [465]:
#Visualizamos el diccionario de datos del dataset principal 
diccionario

Var_Id                                          Var_Title  \
0        1                                          ID_CLIENT   
1        2                                         CLERK_TYPE   
2        3                                        PAYMENT_DAY   
3        4                        APPLICATION_SUBMISSION_TYPE   
4        5  QUANT_ADDITIONAL_CARDS/QUANT_ADDITIONAL_CARDS_...   
5        6                                POSTAL_ADDRESS_TYPE   
6        7                                                SEX   
7        8                                     MARITAL_STATUS   
8        9                                   QUANT_DEPENDANTS   
9       10                                    EDUCATION_LEVEL   
10      11                                     STATE_OF_BIRTH   
11      12                                      CITY_OF_BIRTH   
12      13                                        NACIONALITY   
13      14                                  RESIDENCIAL_STATE   
14      15                                   RESIDENCIAL_CITY   
15      16                                RESIDENCIAL_BOROUGH   
16      17                             FLAG_RESIDENCIAL_PHONE   
17      18                        RESIDENCIAL_PHONE_AREA_CODE   
18      19                                     RESIDENCE_TYPE   
19      20                                MONTHS_IN_RESIDENCE   
20      21                                  FLAG_MOBILE_PHONE   
21      22                                         FLAG_EMAIL   
22      23        PERSONAL_MONTHLY_INCOME/PERSONAL_NET_INCOME   
23      24                                      OTHER_INCOMES   
24      25                                          FLAG_VISA   
25      26                                    FLAG_MASTERCARD   
26      27                                        FLAG_DINERS   
27      28                              FLAG_AMERICAN_EXPRESS   
28      29                                   FLAG_OTHER_CARDS   
29      30                             QUANT_BANKING_ACCOUNTS   
30      31                     QUANT_SPECIAL_BANKING_ACCOUNTS   
31      32                              PERSONAL_ASSETS_VALUE   
32      33                                         QUANT_CARS   
33      34                                            COMPANY   
34      35                                 PROFESSIONAL_STATE   
35      36                                  PROFESSIONAL_CITY   
36      37                               PROFESSIONAL_BOROUGH   
37      38                            FLAG_PROFESSIONAL_PHONE   
38      39                       PROFESSIONAL_PHONE_AREA_CODE   
39      40                                  MONTHS_IN_THE_JOB   
40      41                                    PROFESSION_CODE   
41      42                                    OCCUPATION_TYPE   
42      43                               MATE_PROFESSION_CODE   
43      44                                    EDUCATION_LEVEL   
44      45                         FLAG_HOME_ADDRESS_DOCUMENT   
45      46                                            FLAG_RG   
46      47                                           FLAG_CPF   
47      48                                  FLAG_INCOME_PROOF   
48      49                                            PRODUCT   
49      50                                   FLAG_ACSP_RECORD   
50      51                                                AGE   
51      52                                  RESIDENCIAL_ZIP_3   
52      53                                 PROFESSIONAL_ZIP_3   
53      54                           TARGET_LABEL_BAD=1 (tgt)   
54      55                                            ID_SHOP   
55      56                   FLAG_RESIDENCE_TOWN=WORKING_TOWN   
56      57                                        MATE_INCOME   
57      58                                 FLAG_CONTACT_PHONE   

                                      Var_Description  \
0   Sequential number for the applicant (to be use...   
1                                        Not informed 

### Ocupaciones

**Contenido**

El segundo dataset contiene las profesiones y los códigos respectivos a estas.

In [466]:
#Visualizamos el dataset principal
df1.head(5)

PROFESSION_CODE                              PROFESSION
0              999  Healthcare Practitioners and Technical
1              950  Healthcare Practitioners and Technical
2               13     Educational Instruction and Library
3              205                              Production
4              703     Educational Instruction and Library

**Tamaño** 

In [467]:
#Visualizamos el tamaño del dataset
df1.shape

(295, 2)

**Formato**

In [468]:
#Visualizamos la información del dataset
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 295 entries, 0 to 294
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PROFESSION_CODE  295 non-null    int64 
 1   PROFESSION       295 non-null    object
dtypes: int64(1), object(1)
memory usage: 4.7+ KB


## Calidad de datos

### Definición

La calidad de datos se refiere al grado en que los datos se ajustan a los criterios establecidos para su uso. Estos criterios de calidad de los datos abarcan aspectos como exactitud, coherencia, actualización, exhaustividad y decirigibilidad. La calidad de los datos es uno de los principales factores en el éxito de la base de datos. Si los datos no cumplen con los criterios de calidad adecuados, la información generada a partir de dichos datos también será maliciosa o inútil. Por lo tanto, es crítica la necesidad de vigilancia y mejora de la calidad de los datos.

### Etiquetado de variables

#### Contenido

Visualizamos el contenido y las columnas de los datasets nuevamente para registrar los siguientes pasos de calidad.

##### Dataset princial

In [469]:
#Visualizamos las columnas contenidas dentro del dataset principal
df.columns

Index(['ID_CLIENT', 'ID_SHOP', 'SEX', 'MARITAL_STATUS', 'AGE',
       'QUANT_DEPENDANTS', 'EDUCATION', 'FLAG_RESIDENCIAL_PHONE',
       'AREA_CODE_RESIDENCIAL_PHONE', 'PAYMENT_DAY', 'SHOP_RANK',
       'RESIDENCE_TYPE', 'MONTHS_IN_RESIDENCE', 'FLAG_MOTHERS_NAME',
       'FLAG_FATHERS_NAME', 'FLAG_RESIDENCE_TOWN=WORKING_TOWN',
       'FLAG_RESIDENCE_STATE=WORKING_STATE', 'MONTHS_IN_THE_JOB',
       'PROFESSION_CODE', 'MATE_INCOME',
       'FLAG_RESIDENCIAL_ADDRESS=POSTAL_ADDRESS', 'FLAG_OTHER_CARD',
       'QUANT_BANKING_ACCOUNTS', 'PERSONAL_REFERENCE_#1',
       'PERSONAL_REFERENCE_#2', 'FLAG_MOBILE_PHONE', 'FLAG_CONTACT_PHONE',
       'PERSONAL_NET_INCOME', 'COD_APPLICATION_BOOTH',
       'QUANT_ADDITIONAL_CARDS_IN_THE_APPLICATION',
       'FLAG_CARD_INSURANCE_OPTION', 'tgt'],
      dtype='object')

In [470]:
#Visualizamos el contenido del datset principal
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51000 entries, 0 to 50999
Data columns (total 32 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   ID_CLIENT                                  51000 non-null  int64  
 1   ID_SHOP                                    51000 non-null  int64  
 2   SEX                                        50997 non-null  object 
 3   MARITAL_STATUS                             51000 non-null  object 
 4   AGE                                        48528 non-null  float64
 5   QUANT_DEPENDANTS                           51000 non-null  int64  
 6   EDUCATION                                  0 non-null      float64
 7   FLAG_RESIDENCIAL_PHONE                     51000 non-null  object 
 8   AREA_CODE_RESIDENCIAL_PHONE                51000 non-null  int64  
 9   PAYMENT_DAY                                51000 non-null  int64  
 10  SHOP_RANK             

In [471]:
#Visualizamos la distribución del contenido por tipos de datos de la columna del dataset principal 
df.dtypes.value_counts()

object     16
int64      13
float64     3
dtype: int64

**Distribución del contenido**

Los tipos de datos contenidos en el dataset se distribuyen de la siguiente manera

- Columnas de tipo Object: 16
- Columnas de tipo int: 13
- Columnas de tipo float: 3

##### Dataset secundario

In [472]:
#Visualizamos las columnas contenidas dentro del dataset principal
df1.columns

Index(['PROFESSION_CODE', 'PROFESSION'], dtype='object')

In [473]:
#Visualizamos el contenido del datset secundario
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 295 entries, 0 to 294
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PROFESSION_CODE  295 non-null    int64 
 1   PROFESSION       295 non-null    object
dtypes: int64(1), object(1)
memory usage: 4.7+ KB


In [474]:
#Visualizamos la distribución del contenido por tipos de datos de la columna del dataset secundario 
df1.dtypes.value_counts()

int64     1
object    1
dtype: int64

**Distribución del contenido**

Los tipos de datos contenidos en el dataset se distribuyen de la siguiente manera

- Columnas de tipo int: 1
- Columnas de tipo object: 1

#### Prefijos

Realizamos el etiquetado de las variables utilizando los siguientes prefijos que utilizaremos para los tipos de variables se describen de la siguiente manera

- **c_**: Variables numericas (discretas y continuas).
- **v_**: Variables categoricas.
- **d_**: Variables tipo fecha.
- **t_**: Variables de texto (comentarios, descripciones, url, etc.).
- **g_**: Variables geograficas.

#### Etiquetado

**Dataset principal**

In [475]:
#Realizamos el etiquetado de las variables del datset principal de acuerdo a los prefijos anteriormente establecidos
c_feats=['ID_CLIENT', 'ID_SHOP', 'AGE', 'QUANT_DEPENDANTS', 'PAYMENT_DAY', 'SHOP_RANK',
        'AREA_CODE_RESIDENCIAL_PHONE', 'MONTHS_IN_RESIDENCE', 'MATE_INCOME', 'PERSONAL_NET_INCOME', 'COD_APPLICATION_BOOTH', 'PROFESSION_CODE',
        'QUANT_ADDITIONAL_CARDS_IN_THE_APPLICATION', 'MONTHS_IN_THE_JOB', 'QUANT_BANKING_ACCOUNTS']
v_feats=['SEX', 'EDUCATION', 'MARITAL_STATUS', 'FLAG_MOTHERS_NAME', 'FLAG_FATHERS_NAME', 'FLAG_RESIDENCE_STATE=WORKING_STATE',
        'FLAG_RESIDENCIAL_ADDRESS=POSTAL_ADDRESS', 'FLAG_OTHER_CARD', 'RESIDENCE_TYPE', 'FLAG_RESIDENCE_TOWN=WORKING_TOWN', 'FLAG_RESIDENCIAL_PHONE', 'FLAG_MOBILE_PHONE', 
        'FLAG_CONTACT_PHONE', 'FLAG_CARD_INSURANCE_OPTION']
t_feats=['PERSONAL_REFERENCE_#1', 'PERSONAL_REFERENCE_#2']

In [476]:
#Aplicamos 
c_feats_new = ["c_" + x for x in c_feats]
v_feats_new = ["v_" + x for x in v_feats]
t_feats_new = ["t_" + x for x in t_feats]

#Renombramos las columnas
df.rename(columns = dict(zip(c_feats, c_feats_new)), inplace = True)
df.rename(columns = dict(zip(v_feats, v_feats_new)), inplace = True)
df.rename(columns = dict(zip(t_feats, t_feats_new)), inplace = True)

In [477]:
#Transformamos nuestras columnas en minisculas para una mejor visualización y manipulación futura
df.columns = df.columns.str.lower()

In [478]:
#Validamos el etiquetado
df.columns

Index(['c_id_client', 'c_id_shop', 'v_sex', 'v_marital_status', 'c_age',
       'c_quant_dependants', 'v_education', 'v_flag_residencial_phone',
       'c_area_code_residencial_phone', 'c_payment_day', 'c_shop_rank',
       'v_residence_type', 'c_months_in_residence', 'v_flag_mothers_name',
       'v_flag_fathers_name', 'v_flag_residence_town=working_town',
       'v_flag_residence_state=working_state', 'c_months_in_the_job',
       'c_profession_code', 'c_mate_income',
       'v_flag_residencial_address=postal_address', 'v_flag_other_card',
       'c_quant_banking_accounts', 't_personal_reference_#1',
       't_personal_reference_#2', 'v_flag_mobile_phone',
       'v_flag_contact_phone', 'c_personal_net_income',
       'c_cod_application_booth',
       'c_quant_additional_cards_in_the_application',
       'v_flag_card_insurance_option', 'tgt'],
      dtype='object')

**Dataset secundario**

In [479]:
#Realizamos el etiquetado de las variables del datset principal de acuerdo a los prefijos anteriormente establecidos
c_feats=['PROFESSION_CODE']
v_feats=['PROFESSION']

In [480]:
#Aplicamos 
c_feats_new=["c_"+x for x in c_feats]
v_feats_new=["v_"+x for x in v_feats]

#Renombramos las columnas
df1.rename(columns=dict(zip(c_feats, c_feats_new)),inplace=True)
df1.rename(columns=dict(zip(v_feats, v_feats_new)),inplace=True)

In [481]:
#Transformamos nuestras columnas en minisculas para una mejor visualización y manipulación futura
df1.columns = df1.columns.str.lower()

In [482]:
#Validamos el etiquetado
df1.columns

Index(['c_profession_code', 'v_profession'], dtype='object')

### Duplicados

#### General

**Dataset principal**

Con el resultado de siguiente código podemos observar la existencia de **5** valores duplicados en todo el dataset

In [483]:
#Total de registros duplicados de forma general
df.duplicated().sum()

5

In [484]:
#Visualizamos los duplicados 
df[df.duplicated()]

c_id_client  c_id_shop v_sex v_marital_status  c_age  \
50152        39582         24     F                C   27.0   
50213        44401         16     F                O   22.0   
50592        17829         19     F                S   39.0   
50938        32014         20     F                S   37.0   
50944        19388         22     M                C   27.0   

       c_quant_dependants  v_education v_flag_residencial_phone  \
50152                   0          NaN                        Y   
50213                   0          NaN                        Y   
50592                   0          NaN                        Y   
50938                   0          NaN                        Y   
50944                   0          NaN                        N   

       c_area_code_residencial_phone  c_payment_day  c_shop_rank  \
50152                             31             12            0   
50213                             31             28            0   
50592                             31              8            0   
50938                             31              8            0   
50944                             50              8            0   

      v_residence_type  c_months_in_residence v_flag_mothers_name  \
50152                P                     12                   Y   
50213                P                     12                   Y   
50592                P                     96                   Y   
50938                P                    432                   Y   
50944                P                     24                   Y   

      v_flag_fathers_name v_flag_residence_town=working_town  \
50152                   Y                                  Y   
50213                   Y                                  N   
50592                   Y                                  Y   
50938                   Y                                  Y   
50944                   Y                                  N   

      v_flag_residence_state=working_state  c_months_in_the_job  \
50152                                    Y                  120   
50213                                    Y                    0   
50592                                    Y                   12   
50938                                    Y                   12   
50944                                    Y                   96   

       c_profession_code  c_mate_income  \
50152                654            0.0   
50213                999            0.0   
50592                 26            0.0   
50938                205            0.0   
50944                514            0.0   

      v_flag_residencial_address=postal_address v_flag_other_card  \
50152                                         Y                 N   
50213                                         Y                 N   
50592                                         Y                 N   
50938                                         Y                 N   
50944                                         Y                 N   

       c_quant_banking_accounts t_personal_reference_#1  \
50152                         0                 GUIOMAR   
50213                         0   MARIA DO SOCORRO**TIA   
50592                         0                CRISTINA   
50938                         0                 CLAUDIA   
50944                         0            MIN EXERCITO   

      t_personal_reference_#2 v_flag_mobile_phone v_flag_contact_phone  \
50152               JAQUELINE                   N                    N   
50213                   LUZIA                   N                    N   
50592                   TANIA                   N                    N   
50938               ROSANGELA                   N                    N   
50944                     NaN                   N                    N   

      c_personal_net_income  c_cod_application_booth  \
50152                 382.0                        0   
50213                 220.0

In [485]:
#Mantenemos el primer elemento de los duplicados
df.drop_duplicates(keep = 'first', inplace = True)

In [486]:
#Visualizamos el dataset principal y observamos que se han eliminado los duplicados
df

c_id_client  c_id_shop v_sex v_marital_status  c_age  \
0                2         15     F                S   18.0   
1                4         12     F                C    NaN   
2                5         16     F                S   28.0   
3                6         24     M                S   26.0   
4                7         55     F                S   22.0   
...            ...        ...   ...              ...    ...   
50995        29197         16     F                S   72.0   
50996        43802         23     F                S   31.0   
50997         1881          4     F                C   51.0   
50998        42857         25     F                C   71.0   
50999        46722          5     F                V   42.0   

       c_quant_dependants  v_education v_flag_residencial_phone  \
0                       0          NaN                        Y   
1                       0          NaN                        N   
2                       0          NaN                        Y   
3                       0          NaN                        N   
4                       0          NaN                        Y   
...                   ...          ...                      ...   
50995                   0          NaN                        Y   
50996                   0          NaN                        Y   
50997                   0          NaN                        Y   
50998                   0          NaN                        Y   
50999                   0          NaN                        Y   

       c_area_code_residencial_phone  c_payment_day  c_shop_rank  \
0                                 31             20            0   
1                                 31             25            0   
2                                 31             25            0   
3                                 31             28            0   
4                                 31             12            0   
...                              ...            ...          ...   
50995                             31             23            0   
50996                             31             18            0   
50997                             31             20            0   
50998                             50             12            0   
50999                             23             18            0   

      v_residence_type  c_months_in_residence v_flag_mothers_name  \
0                    P                    216                   Y   
1                    P                    180                   Y   
2                                          12                   Y   
3                    P                    180                   Y   
4                    A                      0                   Y   
...                ...                    ...                 ...   
50995                C                    120                   Y   
50996                P                    240                   Y   
50997                P                     12                   Y   
50998                P                    300                   Y   
50999                P                     48                   Y   

      v_flag_fathers_name v_flag_residence_town=working_town  \
0                       Y                                  Y   
1                       Y                                  N   
2                       Y                                  Y   
3                       Y                                  N   
4                       Y                                  Y   
...                   ...                                ...   
50995                   Y                                  Y   
50996                   Y                                  N   
50997                   Y                                  N   
50998                   Y                                  Y   
50999                   Y                                  N   

      v_flag_residence_state=w

**Dataset secundario**

Con el resultado de siguiente código podemos observar la existencia de **0** valores duplicados en todo el dataset.

In [487]:
#Total de registros duplicados de forma general
df1.duplicated().sum()

0

In [488]:
#Visualizamos el dataset principal y observamos que se han eliminado los duplicados
df1

c_profession_code                            v_profession
0                  999  Healthcare Practitioners and Technical
1                  950  Healthcare Practitioners and Technical
2                   13     Educational Instruction and Library
3                  205                              Production
4                  703     Educational Instruction and Library
..                 ...                                     ...
290                152                              Production
291                516      Life, Physical, and Social Science
292                 50     Educational Instruction and Library
293                850      Life, Physical, and Social Science
294                726     Educational Instruction and Library

[295 rows x 2 columns]

### Completitud

#### Dataset principal

In [489]:
df.isnull().sum(0)

c_id_client                                        0
c_id_shop                                          0
v_sex                                              3
v_marital_status                                   0
c_age                                           2472
c_quant_dependants                                 0
v_education                                    50995
v_flag_residencial_phone                           0
c_area_code_residencial_phone                      0
c_payment_day                                      0
c_shop_rank                                        0
v_residence_type                                   0
c_months_in_residence                              0
v_flag_mothers_name                                0
v_flag_fathers_name                                0
v_flag_residence_town=working_town                 0
v_flag_residence_state=working_state               0
c_months_in_the_job                                0
c_profession_code                             

In [490]:
#Revisamos la completitud de las columnas del dataset
completitud(df)

columna  total  completitud
0                                   v_education  50995     0.000000
1                       t_personal_reference_#1  20625    59.554858
2                       t_personal_reference_#2  13886    72.769879
3                                         c_age   2472    95.152466
4                          v_flag_contact_phone   1279    97.491911
5                                 c_mate_income    306    99.399941
6                                         v_sex      3    99.994117
7                                   c_id_client      0   100.000000
8     v_flag_residencial_address=postal_address      0   100.000000
9                             v_flag_other_card      0   100.000000
10                     c_quant_banking_accounts      0   100.000000
11                          v_flag_mobile_phone      0   100.000000
12                        c_personal_net_income      0   100.000000
13                      c_cod_application_booth      0   100.000000
14  c_quant_additional_cards_in_the_application      0   100.000000
15                            c_profession_code      0   100.000000
16                          c_months_in_the_job      0   100.000000
17           v_flag_residence_town=working_town      0   100.000000
18                 v_flag_card_insurance_option      0   100.000000
19                          v_flag_fathers_name      0   100.000000
20                          v_flag_mothers_name      0   100.000000
21                        c_months_in_residence      0   100.000000
22                             v_residence_type      0   100.000000
23                                  c_shop_rank      0   100.000000
24                                c_payment_day      0   100.000000
25                c_area_code_residencial_phone      0   100.000000
26                     v_flag_residencial_phone      0   100.000000
27                           c_quant_dependants      0   100.000000
28                             v_marital_status      0   100.000000
29                                    c_id_shop      0   100.000000
30         v_flag_residence_state=working_state      0   100.000000
31                                          tgt      0   100.000000

**Observaciones**

Se recomienda que las variables esten completas en al menos 80%, de modo que las que no cumplan con esta condición serán eliminadas. En este sentido, y para el presente análisis, las columnas que no cumplen con dicha regla y que deberían ser eliminadas se enlistan a continuación:

- v_education
- t_personal_reference_#1
- t_personal_reference_#2


In [491]:
#Revisamos nuevamente el tamaño antes de eliminar las columnas con información incompleta 
df.shape

(50995, 32)

In [492]:
#Eliminamos las columnas con información incompleta
df.drop(columns = ["v_education", "t_personal_reference_#1", "t_personal_reference_#2"], inplace = True)

In [493]:
##Revisamos nuevamente el tamaño despúes de haber de eliminado las columnas con información incompleta 
df.shape

(50995, 29)

In [494]:
#Revisamos que no quedan columnas incompletas, es decir, con información menor del 80% 
completitud(df)

columna  total  completitud
0                                         c_age   2472    95.152466
1                          v_flag_contact_phone   1279    97.491911
2                                 c_mate_income    306    99.399941
3                                         v_sex      3    99.994117
4                                   c_id_client      0   100.000000
5   c_quant_additional_cards_in_the_application      0   100.000000
6                       c_cod_application_booth      0   100.000000
7                         c_personal_net_income      0   100.000000
8                           v_flag_mobile_phone      0   100.000000
9                      c_quant_banking_accounts      0   100.000000
10                            v_flag_other_card      0   100.000000
11    v_flag_residencial_address=postal_address      0   100.000000
12                            c_profession_code      0   100.000000
13                          c_months_in_the_job      0   100.000000
14         v_flag_residence_state=working_state      0   100.000000
15           v_flag_residence_town=working_town      0   100.000000
16                          v_flag_fathers_name      0   100.000000
17                          v_flag_mothers_name      0   100.000000
18                        c_months_in_residence      0   100.000000
19                             v_residence_type      0   100.000000
20                                  c_shop_rank      0   100.000000
21                                c_payment_day      0   100.000000
22                c_area_code_residencial_phone      0   100.000000
23                     v_flag_residencial_phone      0   100.000000
24                           c_quant_dependants      0   100.000000
25                             v_marital_status      0   100.000000
26                                    c_id_shop      0   100.000000
27                 v_flag_card_insurance_option      0   100.000000
28                                          tgt      0   100.000000

#### Dataset secundario

In [495]:
df1.isnull().sum(0)

c_profession_code    0
v_profession         0
dtype: int64

In [496]:
#Revisamos la completitud de las columnas del dataset
completitud(df1)

columna  total  completitud
0  c_profession_code      0        100.0
1       v_profession      0        100.0

**Observaciones**

No se encuentran variables incompletas


### Consistencia

#### Categóricas

In [497]:
#Revisamos las variables categoricas con la siguiente iteración donde se filtran los valores de las variable de forma única y en un formato de cadena
for i in df.filter(like="v_"):
  print(i)
  values = df[i].astype(str).unique()
  values.sort()
  display(values)
  print("\n")

v_sex


array(['F', 'M', 'n.a', 'nan'], dtype=object)



v_marital_status


array(['C', 'C ', 'D', 'D ', 'O', 'O ', 'S', 'S ', 'V', 'V '],
      dtype=object)



v_flag_residencial_phone


array(['N', 'Y', 'n', 'y'], dtype=object)



v_residence_type


array([' ', ' a', ' c', ' o', ' p', 'A', 'A ', 'C', 'C ', 'O', 'O ', 'P',
       'P ', 'a', 'c', 'o', 'p', 'p '], dtype=object)



v_flag_mothers_name


array(['N', 'Y'], dtype=object)



v_flag_fathers_name


array(['N', 'Y'], dtype=object)



v_flag_residence_town=working_town


array(['N', 'Y'], dtype=object)



v_flag_residence_state=working_state


array(['N', 'Y'], dtype=object)



v_flag_residencial_address=postal_address


array(['N', 'Y'], dtype=object)



v_flag_other_card


array(['N'], dtype=object)



v_flag_mobile_phone


array(['N'], dtype=object)



v_flag_contact_phone


array(['N', 'nan'], dtype=object)



v_flag_card_insurance_option


array(['N'], dtype=object)

**Variables con inconstistencias**

Al revisar las variables categóricas se pueden observar las siguientes variables con inconsistencias y sus respectivas causas:

1. **v_sex**: Esta variable contiene dos formas diferentes de representación de nulos.
2. **v_marital_status**: Esta variable contiene diversas formas de representar cada una de sus categorías debido a que hay espacios antes o despues de la cadena.
3. **v_flag_residencial_phone**: Esta variable contiene diversas formas de representar cada una de sus categorías debido a que hay minusculas y mayusculas para cada una.
4. **v_residence_type**: Esta variable contiene diversas formas de representar cada una de sus categorías debido a que hay minusculas, mayusculas y espacios al inicio o al final, así como nuelos en cada una.

**Observaciones**

Las variables **v_marital_status** y **v_flag_residencial_phone** deberán ser limpiadas más adelante.

#### Continuas

In [498]:
#Revisamos las variables continuas con la siguiente iteración donde se filtran los valores de las variable de forma única y en un formato de cadena
for i in df.filter(like="c_"):
  print(i)
  values = df[i].astype(str).unique()
  values.sort()
  display(values)
  print("\n")

c_id_client


array(['1', '10', '100', ..., '9997', '9998', '9999'], dtype=object)



c_id_shop


array(['1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19',
       '2', '20', '21', '22', '23', '24', '25', '3', '4', '5', '50', '55',
       '6', '66', '7', '77', '8', '81', '9', '96'], dtype=object)



c_age


array(['114.0', '120.0', '126.0', '132.0', '138.0', '144.0', '15.0',
       '16.0', '168.0', '17.0', '18.0', '186.0', '19.0', '192.0', '198.0',
       '20.0', '21.0', '210.0', '22.0', '222.0', '23.0', '24.0', '240.0',
       '246.0', '25.0', '26.0', '264.0', '27.0', '28.0', '282.0', '29.0',
       '30.0', '300.0', '31.0', '318.0', '32.0', '33.0', '34.0', '35.0',
       '354.0', '36.0', '37.0', '372.0', '38.0', '384.0', '39.0', '40.0',
       '41.0', '42.0', '43.0', '44.0', '45.0', '46.0', '47.0', '48.0',
       '49.0', '50.0', '51.0', '52.0', '53.0', '54.0', '55.0', '56.0',
       '57.0', '58.0', '59.0', '60.0', '61.0', '62.0', '63.0', '64.0',
       '65.0', '66.0', '67.0', '68.0', '69.0', '70.0', '71.0', '72.0',
       '73.0', '74.0', '75.0', '76.0', '77.0', '78.0', '79.0', '80.0',
       '81.0', '82.0', '83.0', '84.0', '85.0', '86.0', '87.0', '88.0',
       '89.0', '90.0', '91.0', '92.0', '93.0', '94.0', '95.0', '96.0',
       '97.0', '98.0', '99.0', 'nan'], dtype=object)



c_quant_dependants


array(['0'], dtype=object)



c_area_code_residencial_phone


array(['1', '10', '11', '12', '13', '14', '15', '17', '18', '19', '2',
       '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '3',
       '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '4',
       '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '5',
       '50', '51', '52', '53', '54', '56', '57', '58', '59', '6', '60',
       '61', '62', '63', '64', '65', '67', '68', '69', '7', '70', '8',
       '9'], dtype=object)



c_payment_day


array(['1', '11', '12', '15', '16', '18', '20', '22', '23', '25', '27',
       '28', '3', '6', '8', '9'], dtype=object)



c_shop_rank


array(['0', '2', '3'], dtype=object)



c_months_in_residence


array(['0', '1020', '108', '1116', '1176', '1188', '12', '120', '132',
       '144', '156', '168', '180', '192', '204', '216', '228', '24',
       '240', '252', '264', '276', '288', '300', '312', '324', '336',
       '348', '36', '360', '372', '384', '396', '408', '420', '432',
       '444', '456', '468', '48', '480', '492', '504', '516', '528',
       '540', '552', '564', '576', '588', '60', '600', '612', '624',
       '636', '648', '660', '672', '684', '696', '708', '72', '720',
       '732', '744', '756', '780', '792', '804', '816', '828', '84',
       '840', '852', '876', '888', '900', '912', '96'], dtype=object)



c_months_in_the_job


array(['0', '108', '1104', '1176', '12', '120', '132', '144', '156',
       '168', '180', '192', '204', '216', '228', '24', '240', '252',
       '264', '276', '288', '300', '312', '324', '336', '348', '36',
       '360', '372', '384', '396', '408', '420', '432', '444', '456',
       '468', '48', '480', '492', '504', '516', '552', '576', '588', '60',
       '600', '612', '660', '684', '708', '72', '780', '84', '852', '96'],
      dtype=object)



c_profession_code


array(['0', '1', '10', '100', '102', '103', '105', '106', '107', '108',
       '109', '11', '110', '113', '114', '115', '116', '117', '118',
       '119', '12', '120', '121', '123', '124', '125', '127', '129', '13',
       '130', '131', '132', '133', '134', '135', '136', '137', '138',
       '139', '140', '141', '15', '150', '151', '152', '153', '154',
       '155', '158', '159', '16', '161', '163', '164', '165', '166',
       '167', '168', '169', '17', '18', '19', '2', '201', '202', '203',
       '205', '209', '21', '210', '211', '213', '214', '215', '217',
       '218', '219', '22', '220', '221', '222', '223', '24', '25', '250',
       '251', '253', '254', '255', '26', '260', '261', '262', '263',
       '264', '265', '266', '28', '29', '3', '30', '300', '301', '302',
       '304', '305', '307', '308', '309', '31', '310', '311', '312', '32',
       '33', '34', '35', '350', '351', '352', '36', '37', '38', '39', '4',
       '40', '400', '401', '402', '403', '41', '42', '43', '44', '45',



c_mate_income


array(['0.0', '1.0', '100.0', '1000.0', '10000.0', '1002.0', '1003.0',
       '1010.0', '1012.0', '1014.0', '1016.0', '1017.0', '1018.0',
       '1020.0', '1021.0', '1022.0', '1025.0', '1026.0', '1030.0',
       '1032.0', '1033.0', '1034.0', '1036.0', '1037.0', '1038.0',
       '1050.0', '1051.0', '1052.0', '1054.0', '1055.0', '1056.0',
       '1057.0', '1060.0', '1064.0', '1065.0', '1066.0', '1069.0',
       '1070.0', '1072.0', '1075.0', '1076.0', '1077.0', '1083.0',
       '1086.0', '1092.0', '1096.0', '1098.0', '1099.0', '1100.0',
       '1108.0', '1109.0', '1110.0', '1120.0', '1121.0', '1126.0',
       '1131.0', '1132.0', '1133.0', '1136.0', '1141.0', '1144.0',
       '1147.0', '11476.0', '1148.0', '1150.0', '1151.0', '1154.0',
       '1157.0', '1158.0', '1162.0', '1163.0', '1167.0', '1170.0',
       '1177.0', '1184.0', '1200.0', '1201.0', '1205.0', '1209.0',
       '1210.0', '1215.0', '1216.0', '1227.0', '1230.0', '1239.0',
       '1249.0', '1250.0', '1263.0', '1269.0', '1273.0', 



c_quant_banking_accounts


array(['0'], dtype=object)



c_personal_net_income


array(['0.0', '1.0', '10.0', ..., '998.0', '999.0', 'n.a.'], dtype=object)



c_cod_application_booth


array(['0'], dtype=object)



c_quant_additional_cards_in_the_application


array(['0', '1', '2', '3'], dtype=object)

##### Validaciones

Debido a que los valores únicos para las variables continuas son más dificiles de observar haremos validaciones extras de modo que se revisaran casos especiales según la naturaleza de las siguientes variables

**c_id_client**

Observamos que la naturaleza de la variable es correcta.

In [499]:
#Revisamos si existen valores que no sean dígitos
df[df["c_id_client"].map(lambda x: sum([y.isalpha() for y in str(x)])>0)]

Empty DataFrame
Columns: [c_id_client, c_id_shop, v_sex, v_marital_status, c_age, c_quant_dependants, v_flag_residencial_phone, c_area_code_residencial_phone, c_payment_day, c_shop_rank, v_residence_type, c_months_in_residence, v_flag_mothers_name, v_flag_fathers_name, v_flag_residence_town=working_town, v_flag_residence_state=working_state, c_months_in_the_job, c_profession_code, c_mate_income, v_flag_residencial_address=postal_address, v_flag_other_card, c_quant_banking_accounts, v_flag_mobile_phone, v_flag_contact_phone, c_personal_net_income, c_cod_application_booth, c_quant_additional_cards_in_the_application, v_flag_card_insurance_option, tgt]
Index: []

**c_age**

A continuacipon se obtuvieron los valores que podrían no ser digitos, y aunque tuvimos resultados a la salida, en realidad solo estamos obteniendo los valores nulos, los **2,472** observados como nulos en la sección de completitud, de modo que se valida que no haya texto en la variable. 

Asimismo, se obtuvieron los registros que contenian edades fuera de lo normalmente establecido por los otorgantes de crédito. Bancos como **BBVA** o **Banco Azteca** han establecido un rango de edad para considerar el otorgamiento de un crédito personal, el cual se define entre los 18 y 85 años. Sin embargo, para este caso considero dejar **15** años más como máximo a considerar (100 años) pues despues de esta edad la mora podría ser más alta y los bancos no participrían en otorgar créditos a este segmento de personas. 

Es importante señalar que esperamos tener ajustes en esta variable en la sección de outliers

In [500]:
#Revisamos si existen valores que no sean dígitos
df[df["c_age"].map(lambda x: sum([y.isalpha() for y in str(x)])>0)]

c_id_client  c_id_shop v_sex v_marital_status  c_age  \
1                4         12     F                C    NaN   
8               11         25     F                S    NaN   
16              21          9     M                C    NaN   
107            131         24     F                S    NaN   
145            176         25     M                C    NaN   
...            ...        ...   ...              ...    ...   
49958        49835         24     F                O    NaN   
49973        49914         25     F                C    NaN   
49988        49963         11     F                C    NaN   
49994        49975         12     M                S    NaN   
50000        32598         23     F                C    NaN   

       c_quant_dependants v_flag_residencial_phone  \
1                       0                        N   
8                       0                        Y   
16                      0                        Y   
107                     0                        Y   
145                     0                        N   
...                   ...                      ...   
49958                   0                        Y   
49973                   0                        Y   
49988                   0                        Y   
49994                   0                        Y   
50000                   0                        Y   

       c_area_code_residencial_phone  c_payment_day  c_shop_rank  \
1                                 31             25            0   
8                                 31             18            0   
16                                 5             18            0   
107                               31             12            0   
145                               50             18            0   
...                              ...            ...          ...   
49958                             31             20            0   
49973                             31             18            0   
49988                             31             20            0   
49994                             31             18            0   
50000                             31             28            0   

      v_residence_type  c_months_in_residence v_flag_mothers_name  \
1                    P                    180                   Y   
8                    P                     60                   Y   
16                   C                    264                   Y   
107                  C                    228                   Y   
145                  P                     36                   Y   
...                ...                    ...                 ...   
49958                P                    600                   Y   
49973                P                     72                   Y   
49988                P                     36                   Y   
49994                P                    216                   Y   
50000                P                     72                   Y   

      v_flag_fathers_name v_flag_residence_town=working_town  \
1                       Y                                  N   
8                       Y                                  N   
16                      Y                                  N   
107                     Y                                  Y   
145                     Y                                  Y   
...                   ...                                ...   
49958                   Y                                  N   
49973                   Y                                  N   
49988                   Y                                  N   
49994                   Y                                  N   
50000                   Y                                  Y   

      v_flag_residence_state=working_state  c_months_in_the_job  \
1                                        Y                   24   
8                                        Y          

In [501]:
#Revisamos el porcentaje que representa la población mayor de 95 años que ha solicitado un crédito, el cual apenas representa un 0.18% de todos los registros
#Podemos convertir a NaN los registros mayores a 95.
(df[df["c_age"] > 95 & (df["c_age"] > 95)].shape[0]/df.shape[0]) * 100

95.15246592803216

In [502]:
df[(df["c_age"] > 85) | (df["c_age"] < 17)]

c_id_client  c_id_shop v_sex v_marital_status  c_age  \
777            965         22     F                O  318.0   
4184          5212         25     F                V  300.0   
5003          6237         22     F                O   86.0   
5236          6533          5     M                S   15.0   
5945          7416         17     F                S  132.0   
...            ...        ...   ...              ...    ...   
50969         4777         24     F                C   94.0   
50971        20761          3     F                S   88.0   
50985         4566         22     F                S   86.0   
50988        35227         15     F                S   88.0   
50993        11178         66     F                S   87.0   

       c_quant_dependants v_flag_residencial_phone  \
777                     0                        Y   
4184                    0                        Y   
5003                    0                        N   
5236                    0                        Y   
5945                    0                        Y   
...                   ...                      ...   
50969                   0                        N   
50971                   0                        N   
50985                   0                        N   
50988                   0                        Y   
50993                   0                        Y   

       c_area_code_residencial_phone  c_payment_day  c_shop_rank  \
777                               31             28            0   
4184                              31             20            0   
5003                              50             28            0   
5236                              24             28            0   
5945                              31             12            0   
...                              ...            ...          ...   
50969                             50             12            0   
50971                             50             12            0   
50985                             50             28            0   
50988                             31             12            0   
50993                             31             25            0   

      v_residence_type  c_months_in_residence v_flag_mothers_name  \
777                  P                    360                   Y   
4184                 A                      0                   Y   
5003                 P                     84                   Y   
5236                 C                    240                   Y   
5945                 P                     60                   Y   
...                ...                    ...                 ...   
50969                P                    216                   Y   
50971                A                     60                   Y   
50985                P                    312                   Y   
50988                P                    240                   Y   
50993                C                    240                   Y   

      v_flag_fathers_name v_flag_residence_town=working_town  \
777                     N                                  N   
4184                    Y                                  N   
5003                    N                                  N   
5236                    Y                                  N   
5945                    Y                                  N   
...                   ...                                ...   
50969                   Y                                  N   
50971                   Y                                  N   
50985                   Y                                  N   
50988                   Y                                  N   
50993                   Y                                  Y   

      v_flag_residence_state=working_state  c_months_in_the_job  \
777                                      Y                  240   
4184                                     Y          

**c_months_in_residence**

Los meses en la vivienda convertidos en años deberían ser estrictamente menores o iguales a los años de vida, pues no puedes vivir más años en una residencia que de los años que tienes cumplidos. Validamos estos casos realizando una resta entre la variable **c_months_in_residence** y la variable **c_age** y confirmando los casos que resuktaran negativos. Debido a que no sabemos cuál es la variable erronea podemos colocar un NaN en ambos casos.

In [503]:
df[(df["c_age"] - df["c_months_in_residence"]/12).map(lambda x:x<0)]

c_id_client  c_id_shop v_sex v_marital_status  c_age  \
31              37         25     M                S   18.0   
44              53          6     F                O   17.0   
172            212          2     M                S   20.0   
201            251         22     F                C   38.0   
214            270         17     M                S   18.0   
...            ...        ...   ...              ...    ...   
50692        11524         15     F                C   32.0   
50768        47582         10     F                S   27.0   
50796        17113         24     F                C   30.0   
50830        14258         24     F                D   24.0   
50890        17946         15     F                S   24.0   

       c_quant_dependants v_flag_residencial_phone  \
31                      0                        Y   
44                      0                        Y   
172                     0                        Y   
201                     0                        Y   
214                     0                        N   
...                   ...                      ...   
50692                   0                        Y   
50768                   0                        Y   
50796                   0                        Y   
50830                   0                        Y   
50890                   0                        Y   

       c_area_code_residencial_phone  c_payment_day  c_shop_rank  \
31                                31             23            0   
44                                34             12            0   
172                               31             28            0   
201                               31             12            0   
214                               50              8            0   
...                              ...            ...          ...   
50692                             31             12            0   
50768                             31             12            0   
50796                             50             18            0   
50830                             31             25            0   
50890                             31             18            0   

      v_residence_type  c_months_in_residence v_flag_mothers_name  \
31                   P                    228                   Y   
44                   P                    216                   Y   
172                  C                    252                   Y   
201                  P                    480                   Y   
214                  C                    228                   Y   
...                ...                    ...                 ...   
50692                P                    756                   Y   
50768                P                    384                   Y   
50796                P                    420                   N   
50830                P                    360                   Y   
50890                P                    456                   Y   

      v_flag_fathers_name v_flag_residence_town=working_town  \
31                      Y                                  N   
44                      Y                                  Y   
172                     Y                                  Y   
201                     Y                                  Y   
214                     Y                                  N   
...                   ...                                ...   
50692                   Y                                  N   
50768                   Y                                  Y   
50796                   Y                                  Y   
50830                   Y                                  Y   
50890                   Y                                  N   

      v_flag_residence_state=working_state  c_months_in_the_job  \
31                                       Y                    0   
44                                       Y          

In [504]:
#Obtenemos el porcentaje de datos invalidos debido a la lógica de años con vida y años de vivienda
(df[(df["c_age"] - df["c_months_in_residence"]/12).map(lambda x:x<0)].shape[0]/df.shape[0]) * 100

1.9158741053044417

**c_months_in_the_job**

Los meses en la vivienda convertidos en años deberían ser estrictamente menores o iguales a los años de vida, pues no puedes vivir más años en una residencia que de los años que tienes cumplidos. Validamos estos casos realizando una resta entre la variable **c_months_in_residence** y la variable **c_age** y confirmando los casos que resuktaran negativos. Debido a que no sabemos cuál es la variable erronea podemos colocar un NaN en ambos casos.

In [505]:
df[(df["c_age"] - df["c_months_in_the_job"]/12).map(lambda x:x<15)]

c_id_client  c_id_shop v_sex v_marital_status  c_age  \
105            129         25     M                S   18.0   
230            292         25     F                S   21.0   
301            377         15     F                C   20.0   
343            424         55     F                S   33.0   
469            585         22     F                S   19.0   
...            ...        ...   ...              ...    ...   
50790         6734         55     M                C   25.0   
50869        16196         55     F                S   33.0   
50887        46947         15     F                D   24.0   
50943        43575         13     M                C   38.0   
50970        28135         55     F                D   36.0   

       c_quant_dependants v_flag_residencial_phone  \
105                     0                        Y   
230                     0                        Y   
301                     0                        Y   
343                     0                        Y   
469                     0                        Y   
...                   ...                      ...   
50790                   0                        Y   
50869                   0                        Y   
50887                   0                        Y   
50943                   0                        Y   
50970                   0                        Y   

       c_area_code_residencial_phone  c_payment_day  c_shop_rank  \
105                               50             20            0   
230                               31              8            0   
301                               31             18            0   
343                               50             12            0   
469                               31             25            0   
...                              ...            ...          ...   
50790                             31             12            0   
50869                             31             18            0   
50887                             31             18            0   
50943                             31              3            0   
50970                             50             20            0   

      v_residence_type  c_months_in_residence v_flag_mothers_name  \
105                  P                    120                   Y   
230                                        24                   Y   
301                                       168                   Y   
343                  P                      0                   Y   
469                  P                    228                   Y   
...                ...                    ...                 ...   
50790                P                    120                   Y   
50869                P                     96                   Y   
50887                P                    180                   Y   
50943                P                     48                   Y   
50970                A                    276                   Y   

      v_flag_fathers_name v_flag_residence_town=working_town  \
105                     Y                                  N   
230                     Y                                  Y   
301                     Y                                  Y   
343                     Y                                  N   
469                     Y                                  N   
...                   ...                                ...   
50790                   Y                                  Y   
50869                   Y                                  N   
50887                   Y                                  Y   
50943                   Y                                  N   
50970                   Y                                  N   

      v_flag_residence_state=working_state  c_months_in_the_job  \
105                                      Y                   60   
230                                      Y          

In [506]:
#Obtenemos el porcentaje de datos invalidos debido a la lógica de años con vida y años de vivienda
(df[(df["c_age"] - df["c_months_in_the_job"]/12).map(lambda x:x<15)].shape[0]/df.shape[0]) * 100

1.2275713305226004

In [507]:
df[(df["c_age"] - df["c_months_in_residence"]/12).map(lambda x:x<0) | (df["c_age"] - df["c_months_in_the_job"]/12).map(lambda x:x<15) | (df["c_age"] > 95)]

c_id_client  c_id_shop v_sex v_marital_status  c_age  \
31              37         25     M                S   18.0   
44              53          6     F                O   17.0   
105            129         25     M                S   18.0   
172            212          2     M                S   20.0   
201            251         22     F                C   38.0   
...            ...        ...   ...              ...    ...   
50894         8031         17     F                C   97.0   
50915        31297         16     F                S   98.0   
50943        43575         13     M                C   38.0   
50962        45428         25     M                C   99.0   
50970        28135         55     F                D   36.0   

       c_quant_dependants v_flag_residencial_phone  \
31                      0                        Y   
44                      0                        Y   
105                     0                        Y   
172                     0                        Y   
201                     0                        Y   
...                   ...                      ...   
50894                   0                        N   
50915                   0                        Y   
50943                   0                        Y   
50962                   0                        Y   
50970                   0                        Y   

       c_area_code_residencial_phone  c_payment_day  c_shop_rank  \
31                                31             23            0   
44                                34             12            0   
105                               50             20            0   
172                               31             28            0   
201                               31             12            0   
...                              ...            ...          ...   
50894                             50              3            0   
50915                             31             12            0   
50943                             31              3            0   
50962                             50             18            0   
50970                             50             20            0   

      v_residence_type  c_months_in_residence v_flag_mothers_name  \
31                   P                    228                   Y   
44                   P                    216                   Y   
105                  P                    120                   Y   
172                  C                    252                   Y   
201                  P                    480                   Y   
...                ...                    ...                 ...   
50894                P                    408                   Y   
50915                P                    384                   Y   
50943                P                     48                   Y   
50962                P                     72                   Y   
50970                A                    276                   Y   

      v_flag_fathers_name v_flag_residence_town=working_town  \
31                      Y                                  N   
44                      Y                                  Y   
105                     Y                                  N   
172                     Y                                  Y   
201                     Y                                  Y   
...                   ...                                ...   
50894                   Y                                  N   
50915                   N                                  N   
50943                   Y                                  N   
50962                   Y                                  Y   
50970                   Y                                  N   

      v_flag_residence_state=working_state  c_months_in_the_job  \
31                                       Y                    0   
44                                       Y          

**Variables con inconstistencias**

Al revisar las variables continuas se pueden observar las siguientes variables con inconsistencias y sus respectivas causas:

1. **c_age**: Esta variable contiene valores con registros de edad fuera de lo común, lógico y esperanza de vida de una persona.
2. **c_months_in_residence**: Hay registros que no coinciden con la condición de ser menores los meses de vivienda que con los meses en residencia.
3. **c_months_in_the_job**: Hay registros que no coinciden con la condición de ser menores los meses de en el trabajo que con los meses en residencia.

**Observaciones**

- La variable **c_personal_net_income** contiene valores NaN, pero parecen no estar bien definidos.
- La variable **c_age** deberá ser modificada en la sección de outliers.
- La variable **c_months_in_residence** deberá ser modificada en la sección de outliers.
- La variable **c_months_in_the_job** deberá ser modificada en la sección de outliers.

#### Transformación a valores NaN

Convertimos las variales con valores nulos incorrectos a valores NaN.

In [508]:
#v_sex
df['v_sex']=df['v_sex'].where(df['v_sex']!='n.a',np.nan)
display(df['v_sex'].value_counts())

#v_residence_type
df['v_residence_type']=df['v_residence_type'].where(df['v_residence_type']!=' ',np.nan)
display(df['v_residence_type'].value_counts())

#c_personal_net_income
df['c_personal_net_income']=df['c_personal_net_income'].where(df['c_personal_net_income']!='n.a.',np.nan)
display(df['c_personal_net_income'].value_counts())

F    34944
M    15257
Name: v_sex, dtype: int64

P     31970
A      5538
C      3783
O      1782
 p      211
P        84
p        50
 a       35
 c       23
A        10
 o        9
a         9
O         4
C         4
c         4
o         2
p         1
Name: v_residence_type, dtype: int64

300.0     2562
500.0     2033
400.0     2006
180.0     1862
600.0     1534
          ... 
2238.0       1
2411.0       1
3077.0       1
2626.0       1
1559.0       1
Name: c_personal_net_income, Length: 2651, dtype: int64

### Completitud

Revisamos de nuevo la completitud

In [509]:
#No tenemos variables con una complitud menor al 80%
completitud(df)

columna  total  completitud
0                              v_residence_type   7476    85.339739
1                                         c_age   2472    95.152466
2                          v_flag_contact_phone   1279    97.491911
3                                         v_sex    794    98.442985
4                                 c_mate_income    306    99.399941
5                         c_personal_net_income    198    99.611727
6                                   c_id_client      0   100.000000
7   c_quant_additional_cards_in_the_application      0   100.000000
8                       c_cod_application_booth      0   100.000000
9                           v_flag_mobile_phone      0   100.000000
10                     c_quant_banking_accounts      0   100.000000
11                            v_flag_other_card      0   100.000000
12    v_flag_residencial_address=postal_address      0   100.000000
13                            c_profession_code      0   100.000000
14                          c_months_in_the_job      0   100.000000
15           v_flag_residence_town=working_town      0   100.000000
16                 v_flag_card_insurance_option      0   100.000000
17                          v_flag_fathers_name      0   100.000000
18                          v_flag_mothers_name      0   100.000000
19                        c_months_in_residence      0   100.000000
20                                  c_shop_rank      0   100.000000
21                                c_payment_day      0   100.000000
22                c_area_code_residencial_phone      0   100.000000
23                     v_flag_residencial_phone      0   100.000000
24                           c_quant_dependants      0   100.000000
25                             v_marital_status      0   100.000000
26                                    c_id_shop      0   100.000000
27         v_flag_residence_state=working_state      0   100.000000
28                                          tgt      0   100.000000

### Limpieza y transformación de variables

In [510]:
#Limpiamos la variable v_marital_status, misma que había sido identificada con anterioridad
df['v_marital_status'].value_counts(1)

S     0.504932
C     0.342543
O     0.055417
V     0.048201
D     0.043024
S     0.003138
C     0.001941
O     0.000294
V     0.000275
D     0.000235
Name: v_marital_status, dtype: float64

In [511]:
#Aplicamos la limpieza y df["v_marital_status"]=df["v_marital_status"].map(clean_text)
df['v_marital_status'].value_counts(1)

s    0.508069
c    0.344485
o    0.055711
v    0.048475
d    0.043259
Name: v_marital_status, dtype: float64

In [512]:
#Limpiamos la variable v_marital_status, misma que había sido identificada con anterioridad
df['v_marital_status'].value_counts(1)

s    0.508069
c    0.344485
o    0.055711
v    0.048475
d    0.043259
Name: v_marital_status, dtype: float64